### PR TITLE
Made correct block for requests with content-type and content-length headers in case with bodyless’s methods

### PR DESCRIPTION
**Pull request by alegz-filomaphitskay**
Contributes to #1297

### CODE REVIEW

File: fw/http_parser.c [hunk 1]
```diff
@@ -3753,6 +3730,39 @@ tfw_http_init_parser_req(TfwHttpReq *req)
...
+			&& IS_BODYLESS_METHOD(req->method_override))
+		    || IS_BODYLESS_METHOD(req->method)) {
+				T_WARN("%s: Content-Length or Content-Type"
+				       " not allowed to be used with such"
+				       " (overridden) method\n", __func__);
+				return T_DROP;
+		}
+	}
+
+	return T_OK;
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:3761]: I'm worried about the return value. Maybe need to use T_POSTPONE?

File: fw/http_parser.c [hunk 2]
```diff
@@ -3753,6 +3730,39 @@ tfw_http_init_parser_req(TfwHttpReq *req)
...
+	 */
+	if (TFW_STR_NOT_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])
+	    || TFW_STR_NOT_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE]))
+	{
+		/* Method override either honored or request message
+		 * with method override header dropped later in processing */
+		if ((req->method_override
+			&& IS_BODYLESS_METHOD(req->method_override))
+		    || IS_BODYLESS_METHOD(req->method)) {
+				T_WARN("%s: Content-Length or Content-Type"
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:3754]: Maybe better to use T_DBG3?

**Comment by ttaym** [comment]
One general comment. I have not found any resources about a prevalence of http override headers in http2. Do you have some links on it? May be that functionality not needed in http2

**Comment by alegz-filomaphitskay** [comment]
> One general comment. I have not found any resources about a prevalence of http override headers in http2. Do you have some links on it?
No, I haven't. But also I don't have any references confirming that the override mechanism is not exist.

I can only quote [the RFC](https://datatracker.ietf.org/doc/html/rfc7540#:~:text=HTTP/2%20is%20intended,those%20semantics%20has%20changed.):
> HTTP/2 is intended to be as compatible as possible with current uses
>    of HTTP.  This means that, from the application perspective, the
>    features of the protocol are largely unchanged.  To achieve this, all
>    request and response semantics are preserved, although the syntax of
>    conveying those semantics has changed.
Can this serve as proof that the override exists?

File: fw/http_parser.c [hunk 3]
```diff
@@ -7646,7 +7647,7 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
 		case TFW_CHAR4_INT('c', 'o', 'n', 'n'):
 			if (unlikely(!__data_available(p, 9)))
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:7648]: Need use ___data_available(p, **10**)_

File: fw/http_parser.c [hunk 4]
```diff
@@ -8340,17 +8341,19 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
 	 * header if more data is found afer 'd'
 	 */
 	__FSM_STATE(Req_HdrX_Http_Metho, cold) {
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:8343]: Pay more attention to this, please.

File: fw/http_parser.c [hunk 5]
```diff
@@ -8340,17 +8341,20 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
...
-		if (c != 'd')
+		if (unlikely(c != 'd'))
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:8344]: The case-insensitive approach is only used with HTTP/1. HTTP/2 processing uses a case-sensitive approach.

File: fw/http_parser.c [hunk 6]
```diff
@@ -3753,6 +3730,39 @@ tfw_http_init_parser_req(TfwHttpReq *req)
...
+				       " not allowed to be used with such"
+				       " (overridden) method\n", __func__);
+				return T_DROP;
+		}
+	}
+
+	return T_OK;
+
+#undef IS_BODYLESS_METHOD
+}
```
[the remark this answers is not in the document]
Comment (ttaym) [reply fw/http_parser.c:3764]: I cannot give you a definitive answer to this question.

File: fw/http_parser.c [hunk 7]
```diff
@@ -8340,17 +8341,20 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
...
 		p += 1;
-		T_DBG3("%s: name fin, h_tag=%d, to=Req_HdrX_Http_Metho len=%lu, "
-		       "off=%lu\n",
-		       __func__, TFW_TAG_HDR_RAW, len, __data_off(p));
-		if (unlikely(__data_off(p) < len))
-			goto RGen_HdrOtherN;
+		T_DBG3("%s: name next, to=Req_HdrX_Http_Method len=%lu,"
+		    " off=%lu\n",
+		    __func__, len, __data_off(p));
+		if (likely(__data_off(p) < len))
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:8351]: This is difficult for me too. But I took the approach from __FSM_H2_NEXT_n.

File: fw/http_parser.c [hunk 8]
```diff
@@ -1193,34 +1193,11 @@ __FSM_STATE(RGen_BodyInit, cold) {					\
...
-		}							\
-		else if (req->method == TFW_HTTP_METH_GET		\
-			 || req->method == TFW_HTTP_METH_HEAD		\
-			 || req->method == TFW_HTTP_METH_DELETE		\
-			 || req->method == TFW_HTTP_METH_TRACE)		\
-		{							\
-			TFW_PARSER_BLOCK(RGen_BodyInit);		\
-		}							\
+									\
+	if (tfw_http_parse_req_on_headers_done(req)) {			\
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:1197]: > tfw_http_parse_req has no even one explicit function call in it. But you introduce one.
What about call of tfw_http_msg_hdr_open(), tfw_http_sess_max_misses(), test_bit(), set_bit()? The tfw_http_parse_req_on_headers_done() is called only once during processing of request.
Do we have some rules about using of function call?

File: fw/http_parser.c (hunk 1 above)
Comment (krizhanovsky) [reply fw/http_parser.c:3761]: We use the `T_*` return value constants to make the processing logic more explicit - what should we do next: read another chunk, pass or block the current chunk. In this function I think basic true/false would be fine as well as `T_OK` or `T_DROP`

File: fw/http_parser.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:3764]: IIRC we only use it for functions using labels as data, i.e. the state machine ones. Anyway there is no reason to annotate this function as using non standard stack

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
Looks good in general, but it seems there are couple of issues and places for cleanups. I'm not sure about the HTTP/2 HEADERS_END vs DATA statement, so probably we need to discuss it on the status call.

File: fw/http_parser.c [hunk 9]
```diff
@@ -8741,10 +8746,9 @@ tfw_h2_parse_req(void *req_data, unsigned char *data, size_t len,
...
-	else if (unlikely(req->method == TFW_HTTP_METH_GET
-		 || req->method == TFW_HTTP_METH_HEAD
-		 || req->method == TFW_HTTP_METH_DELETE
-		 || req->method == TFW_HTTP_METH_TRACE))
+	else if ((req->method_override
+			&& TFW_HTTP_IS_METH_BODYLESS(req->method_override))
+		    || TFW_HTTP_IS_METH_BODYLESS(req->method))
 	{
 		r = T_DROP;
 	}
```
Comment (krizhanovsky) [inline fw/http_parser.c:8754]: Could you please adjust coding style of the block as
```C
        if (likely(type != HTTP2_DATA)) {
                r = tfw_hpack_decode(&ctx->hpack, data, len, req, parsed);
        }
        else if ((req->method_override
                  && TFW_HTTP_IS_METH_BODYLESS(req->method_override))
                 || TFW_HTTP_IS_METH_BODYLESS(req->method))
        {
                r = T_DROP;
        }
        else {
                r = tfw_h2_parse_body(data, len, req, parsed);
        }
```

**Review by krizhanovsky — COMMENTED** [review]
There are either some comments remain not fixed or there are no issues. Let's discuss the comments on the status call.

File: fw/http_parser.c [hunk 10]
```diff
@@ -8340,17 +8341,20 @@ tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
...
-		T_DBG3("%s: name fin, h_tag=%d, to=Req_HdrX_Http_Metho len=%lu, "
-		       "off=%lu\n",
-		       __func__, TFW_TAG_HDR_RAW, len, __data_off(p));
-		if (unlikely(__data_off(p) < len))
-			goto RGen_HdrOtherN;
+		T_DBG3("%s: name next, to=Req_HdrX_Http_Method len=%lu,"
+		    " off=%lu\n",
+		    __func__, len, __data_off(p));
+		if (likely(__data_off(p) < len))
+			__FSM_JMP(Req_HdrX_Http_Method);
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:8353]: Screenshot 2022-03-24 at 08 26 20
I see that my comment is pending. Problem with this.

File: fw/http_parser.c (hunk 9 above)
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:8755]: Done

File: fw/http_parser.c [hunk 11]
```diff
@@ -8767,6 +8771,9 @@ tfw_h2_parse_req_finish(TfwHttpReq *req)
 	if (WARN_ON_ONCE(!tfw_h2_stream_req_complete(req->stream)))
 		return T_DROP;
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:8774]: Done

File: fw/http.c [hunk 12]
```diff
@@ -5303,24 +5303,22 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, struct sk_buff *skb)
...
 					"Request parsing inconsistency");
 				return TFW_BLOCK;
 			}
-			ctx = tfw_h2_context(conn);
-			if (ctx->hdr.flags & HTTP2_F_END_HEADERS) {
-				__set_bit(TFW_HTTP_B_HEADERS_PARSED, req->flags);
-
-				if (unlikely(tfw_http_parse_req_on_headers_done(req)))
-					return TFW_BLOCK;
-			}
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http.c]: > Also, if possible, please write a good commit message what was changed and why, "After code-review" won't be helpful in commit history if one tries to understand the code evolution.
Sure, no problem. But don't we lose the commits history if we do merge with squash commits?

**Pull request merged**